Optimize: 150s => 10s.

htmldoug opened the pull request:
## Summary
Decreases `checkDuplicatesTest` runtime from 150s to 10s by avoiding unnecessary filesystem accesses and hashing.

Added a happy-path unit test for the things I changed.

## Problem
`checkDuplicatesTest` takes roughly 150 seconds to complete on my 27-project build.

CPU usage is quite high:
![checkDuplicatesTest uses maxes out all 12 cpu cores](https://user-images.githubusercontent.com/663139/87760303-89186a80-c7dd-11ea-8325-00ce5d55cdb0.png)

## Other Things I Tried

### Attempt 1: SHA-256 => Murmur
Initially, I suspected the SHA-256 hashing. What else could be burning through so much CPU time?
![1 0 0 hashing](https://user-images.githubusercontent.com/663139/87760693-2378ae00-c7de-11ea-8035-d20fd0f19848.png)

I tried switching to scala's built in Murmur hash function, but it only cut the time down to ~110 seconds.

### Attempt 2: Caching
I figured with 27 projects, there had to be some overlap in my classpath. I dumped `fullClasspath` in each of my projects then flattened:

```
cat **/fullClasspath.txt > all.txt
$ wc -l all.txt
    5495 all.txt
$ cat all.txt | sort | uniq | wc -l
     533
```

So `checkDuplicatesTest` is opening and rehashing each file ~10x. If we had some way to cache the results, we could eliminate this work. But how?

I messed around with a `ConcurrentHashMap`, but couldn't find a good signal to purge it to avoid leaking memory across repeated runs.

I read through [sbt Caching](https://www.scala-sbt.org/1.x/docs/Caching.html), but a few things turned me away.
1. Refactoring - seems like it wants to live at the task level rather than the pure scala level. I didn't want to refactor too much.
2. Filesystem load - looks like it writes to the filesystem. One write per class file seems heavy.

Ultimately, I abandoned the caching approach, since it seems non-trivial to do well.

### Attempt 3: Avoid FS access & hashing
Thinking about FS load, I realized that half of the CPU time is spent in system calls. Is this FS activity?
![1 0 0 system](https://user-images.githubusercontent.com/663139/87760661-1360ce80-c7de-11ea-8f31-43f6d8812f20.png)

Looking deeper, I realized that we're reading and hashing everything, but we only need the hashes if there are multiple files with the same FQCN. Many of these names are unique. We don't need to read or hash these files at all!

I refactored to defer reading the files until after we check if the FQCN is duplicated. I expect that this change was responsible for most of the speedup.

I also spotted that we have a perfectly good CRC32 pre-calculated in the jar/zip central directory. We can grab it for free while traversing the zip entries for the names.

For classes in directories (i.e. outside of jars), we have to compute the CRC32 ourselves, but only if needed on FQCN conflict. I expect the ratio of jars to class files is probably high.

There's some debate on https://stackoverflow.com/q/10953958 about how good CRC32 is as a hash function. (I assume CRC32 and CRC32C perform similarly). It's clearly not as strong as SHA-256, but it also doesn't have to be. On collision, we can just read the whole file contents. Good enough.

### Attempt 4: Back to SHA-256
Thinking further, the CRC32 comparison adds less value than I thought. It only saves IO for the ConflictState.ContentDiffer case. That's the exceptional case rather than the norm since we're probably running checkDuplicatesTest in a CI build. Over the common case, the crc32s will collide and we'll have to go to disk anyway.

checksums.combinations(2).map{ c1 == c2 } perf could regress pathologically badly after my changes in a scenario with lots of duplicates with equal content.

We could replace the combinations with a `checksums.groupBy(_.bytes)` to get back to O(n) disk reads. Simple, but not memory-safe. To get memory safety back, we could either go back to a wide SHA (simple and probabilistically correct) or stream the content and compare bytewise (more complex, faster, 100% correct).

I ended up tossing the CRC32 and bytewise comparison and added back the original SHA-256. Seems like a good balance.

## CLA
I don't see Lightbend CLA checks here, but FWIW, I've signed it previously. You can verify against my akka/akka PRs.

pdalpra commented:
Thanks for this very detailed PR, the work you've put into it is impressive :astonished: 

I'll have a deeper look ASAP, and handle the few upgrades that did not work (as you can see, there has been little maintenance from me on this plugin...)

> I don't see Lightbend CLA checks here, but FWIW, I've signed it previously. You can verify against my akka/akka PRs.

There is no need for the CLA is this case, the plugin is hosted in the `sbt` organization but is not a Lightbend product.

htmldoug commented:
Thanks for the quick response! I've rebased onto master and fixed the first CI failure. Looks like `sbt.io` moved around. I'll fix that next.

> there has been little maintenance from me on this plugin

Hey, it's simple and does the job. Not much is needed. :)

pdalpra commented:
FYI, I just updated sbt to 1.3.13 in #76.

htmldoug commented:
Thanks for the heads up. Should be all rebased and passing. I'm done force pushing for now. Headed to bed. Feel free to change anything to your liking.

htmldoug commented on src/main/scala/com/github/sbt/duplicates/DuplicatesFinderPlugin.scala:
```diff
@@ -71,7 +71,7 @@ object DuplicatesFinderPlugin extends AutoPlugin {
 
     Seq(classpath.classesDuplicates, classpath.resourcesDuplicates)
-      .map(_.filter(conflict => !(reportIfSameContent && conflict.conflictState == ConflictState.ContentEqual)))
+      .map(_.filter(conflict => conflict.conflictState == ConflictState.ContentDiffer || reportIfSameContent))
```
I wasn't sure about this change. The negation of `reportIfSameContent` seemed like a bug. I could be wrong. Double check me on this.

pdalpra replied:
The behaviour stays the same (eg. report conflict if content differs or if content is equal, if it's been enabled), but yours is clearer, let's keep it :+1:

htmldoug replied:
This line is melting my brain. I tried to make a truth table manually to prove I wasn't crazy, but I screwed that up, too--so I wrote a script.

```scala
for {
  r <- Seq(true, false)
  c <- Seq("Equal", "Diff")
} {
  println(s"| $r | $c | ${!(r && c == "Equal")} | ${r || c == "Diff"} |")
}
```


| report | ConflictState | !(r && Equal) | r or Diff |
| -- | -- | -- | -- |
| true | Equal | `false` | `true` |
| true | Diff | true | true |
| false | Equal | `true` | `false️` |
| false | Diff | true | true |

htmldoug replied:
I just guilted myself into writing a scripted test. That's the right solution to settle this. One sec.

htmldoug replied:
Okay, reverting this line causes the new scripted test to fail.
```
[info] [error] (equal/compile:checkDuplicatesTest) Detected 1 classes conflicts:
[info] [error]
[info] [error] - ConflictingFile.class: same content
[info] [error] 	 - /private/var/folders/k2/8q7nrzn13yl4byhlt8yb4mp00000gp/T/sbt_74e4bbb6/sanity/core/target/scala-2.11/classes
[info] [error] 	 - /private/var/folders/k2/8q7nrzn13yl4byhlt8yb4mp00000gp/T/sbt_74e4bbb6/sanity/equal/target/scala-2.11/classes
[info] [error] Total time: 0 s, completed Jul 19, 2020 1:34:12 PM
[error] x sanity / sanity
[error]    {line 25}  Command failed: equal/checkDuplicatesTest failed
```

I'm satisfied now that this change is correct.

htmldoug commented on src/main/scala/com/github/sbt/duplicates/ClasspathEntity.scala:
```diff
@@ -0,0 +1,81 @@
+    obj match {
+      case null => false
+      case other: ClasspathEntity =>
+        name == other.name && hashCode == other.hashCode && bytes.sameElements(other.bytes)
```
Thinking further, the CRC32 comparison is adding less value than I thought. It only saves IO for the `ConflictState.ContentDiffer` case. That's the exceptional case rather than the norm since we're probably running `checkDuplicatesTest` in a CI build. Over the common case, the crc32s will collide and we'll have to go to disk anyway.

[checksums.combinations(2).map{ c1 == c2 }](https://github.com/sbt/sbt-duplicates-finder/blob/3ba841dd40a0f61c1e8cf486b1708f305d54d742/src/main/scala/com/github/sbt/duplicates/Classpath.scala#L30-L32) perf could regress pathologically badly after my changes in a scenario with lots of duplicates with equal content.

I'm thinking we could replace the combinations with a `checksums.groupBy(_.bytes)` to get back to `O(n)` disk reads. Simple, but not memory-safe. To get memory safety back, we could either go back to a wide SHA (simple and probabilistically correct) or stream the content and compare bytewise (more complex, faster, 100% correct).

I'll play with this a bit more. Feedback welcome if you have a strong preference.

htmldoug replied:
I tossed the CRC32 and bytewise comparison and added back the original SHA-256. Seems like a good balance.

I may try to simplify further after dinner.

htmldoug replied:
Bug here. Should be comparing on `sha256`, not `hashCode`. I'll fix.

htmldoug commented:
Ready for review. I'm done for now. Should be in good shape. Happy to make any changes.

pdalpra commented on build.sbt:
```diff
@@ -14,3 +14,14 @@ scalacOptions := Seq(
+}
+testFrameworks += new TestFramework("utest.runner.Framework")
+
+Global / onChangedBuildSource := ReloadOnSourceChanges
```
nitpicking: this is the kind of setting that, IMHO, is usually best left as a per-developer setting (eg. in a `~/.sbt/1.0/config.sbt` file) than a per-project setting.

htmldoug replied:
Sbt 1.2 fails with it. Is there a way to enable it globally just for 1.3+ builds?

pdalpra replied:
sbt's version is exposed though the `sbtVersion` setting, from here you can enable it conditionally.

htmldoug replied:
```
if (sbtVersion.value.split('.')(1).toInt >= 3) {
  Global / onChangedBuildSource := ReloadOnSourceChanges
}
```
```
/Users/doug.roper/.sbt/1.0/config.sbt:2: error: not found: value onChangedBuildSource
  Global / onChangedBuildSource := ReloadOnSourceChanges
           ^
/Users/doug.roper/.sbt/1.0/config.sbt:2: error: not found: value ReloadOnSourceChanges
  Global / onChangedBuildSource := ReloadOnSourceChanges
                                   ^
/Users/doug.roper/.sbt/1.0/config.sbt:1: error: type mismatch;
 found   : Unit
 required: sbt.internal.DslEntry
if (sbtVersion.value.split('.')(1).toInt >= 3) {
^
```

`onChangedBuildSource` and `ReloadOnSourceChanges` just aren't present in the sbt [1.0.0,1.3.0) jars.

pdalpra replied:
Almost forgot I did it, but I borrowed a while ago a snippet of code to do that, in a compatible way: https://github.com/pdalpra/jvm-based-vcsh/blob/master/.sbt/1.0/auto-reload.sbt

htmldoug replied:
Wow, that's quite a best practice. Removed and pushed.

I've been headed down the same path for the past 2h but got stuck at the `Manifest[AnyRef]` and `OptJsonWriter`. Didn't even occur to me that they're unused. Thanks for the rescue. Slightly more satisfying than giving up!

pdalpra commented on src/main/scala/com/github/sbt/duplicates/ClasspathEntity.scala:
```diff
@@ -0,0 +1,81 @@
+
+/**
+  * Lazy & comparable handle to a class or resource within a zip file or directory.
+  * Caches the CRC32.
```
Could be changed to `Caches the file's checksum`, WDYT ? (so that it doesn't require changing the comment if the algorithm changes)

htmldoug replied:
will do!

pdalpra commented on src/main/scala/com/github/sbt/duplicates/ClasspathEntity.scala:
```diff
@@ -0,0 +1,81 @@
+    */
+  protected def withInputStream[T](f: InputStream => T): T
+
+  lazy val sha256: Seq[Byte] =
+    withInputStream { in =>
+      val dis = new DigestInputStream(in, MessageDigest.getInstance("SHA-256"))
+      IO.transfer(dis, NullOutputStream)
+      dis.getMessageDigest.digest().toSeq
+    }
```
Relying on sbt's `IO` helpers, it could be simplified with:

```suggestion
  lazy val sha256: Seq[Byte] = {
    val bytes = withInputStream(Compat.IO.readBytes)
    MessageDigest.getInstance("SHA-256").digest(bytes).toSeq
  }
```

htmldoug replied:
Yours is simpler for sure, but not completely equivalent.

One difference is that `readBytes` pulls the whole content into memory. The 20MB datadog jar is the biggest on my classpath and this section is run for my 27 projects in parallel. It's a bit of a stretch for my project, but that does add up to half of the default sbt heap over the worst case. Given there's no limit to the number of projects or jar size that this plugin may encounter, it seemed worth the extra lines to never have to worry about it again. WDYT? If you're still not convinced, I'll change.

If only we had an `inputStream.foldLeft(digest)((digest, chunk: Array[Byte]) => )`, the memory-safe solution would feel more natural.

pdalpra replied:
Indeed, completely missed that yours was lazily loading, which is indeed better, we can never know how fat an analyzed JAR would be.
Small suggestion though, to avoid using IO and introducing an OutputStream (reads all bytes and until the input is exhausted):

```suggestion
  lazy val sha256: Seq[Byte] =
    withInputStream { in =>
      val dis = new DigestInputStream(in, MessageDigest.getInstance("SHA-256"))
      while(dis.read() != -1) {}
      dis.getMessageDigest.digest().toSeq
    }
```

htmldoug replied:
Can do. I'll keep the `read(arr)` approach from `IO.transfer()`, since it's ~7x faster.

```
class DigestBench {

  private val bytes = Array.iterate[Byte](0.toByte, 8192 * 2)(b => (b + 1).toByte)
  private def stream() = new DigestInputStream(new BufferedInputStream(new ByteArrayInputStream(bytes)), MessageDigest.getInstance("SHA-256"))

  @Benchmark
  def readArr = {
    val in = stream()
    val buf = Array.ofDim[Byte](8192)
    while (in.read(buf) != -1) {}
    in.getMessageDigest.digest()
  }

  @Benchmark
  def readByte = {
    val in = stream()
    while (in.read() != -1) {}
    in.getMessageDigest.digest()
  }
}
```
```
[info] Benchmark              Mode  Cnt      Score       Error  Units
[info] DigestBench.readArr   thrpt    3  24096.342 ± 16817.796  ops/s
[info] DigestBench.readByte  thrpt    3   3458.116 ±   521.676  ops/s
```

pdalpra replied:
Good to know, that's quite a speedup.

pdalpra commented on src/main/scala/com/github/sbt/duplicates/ClasspathEntity.scala:
```diff
@@ -0,0 +1,81 @@
+
+import sbt._
+
+import scala.tools.nsc.interpreter.InputStream
```
Best not import classes from the compiler ;)

htmldoug replied:
oof. auto-import strikes again.

pdalpra commented on src/main/scala/com/github/sbt/duplicates/NullOutputStream.scala:
```diff
@@ -0,0 +1,10 @@
+
+import scala.tools.nsc.interpreter.OutputStream
+
+private[duplicates] object NullOutputStream extends OutputStream {
```
Not required anymore with the change suggested above

htmldoug commented:
I think I've addressed all comments now. Thanks for the thoughtful review and for teaching me some new sbt protips.

htmldoug commented:
Squashed. Thanks!

pdalpra commented:
Just released sbt-duplicates-finder 1.1.0, it should be synced to the sbt mirror soon (you can check when it becomes available here: https://bintray.com/sbt/sbt-plugin-releases)

htmldoug commented:
Thanks so much!
